pkg/tsdb/prometheus/healthcheck.go

```diff
@@ -0,0 +1,93 @@
+package prometheus
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/kindsys"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/tsdb/prometheus/kinds/dataquery"
+	"github.com/grafana/grafana/pkg/tsdb/prometheus/models"
+)
+
+const (
+	refID = "__healthcheck__"
+)
+
+var logger log.Logger = log.New("tsdb.prometheus")
+
+func (s *Service) CheckHealth(ctx context.Context, req *backend.CheckHealthRequest) (*backend.CheckHealthResult,
+	error) {
+	logger := logger.FromContext(ctx)
+	ds, err := s.getInstance(req.PluginContext)
+
+	// check that the datasource exists
+	if err != nil {
+		return getHealthCheckMessage(logger, "error getting datasource info", err)
+	}
+
+	if ds == nil {
+		return getHealthCheckMessage(logger, "", errors.New("invalid datasource info received"))
+	}
+
+	return healthcheck(ctx, req, ds)
+}
+
+func healthcheck(ctx context.Context, req *backend.CheckHealthRequest, i *instance) (*backend.CheckHealthResult, error) {
+	qm := models.QueryModel{
+		LegendFormat: "",
+		UtcOffsetSec: 0,
+		PrometheusDataQuery: dataquery.PrometheusDataQuery{
+			Expr:    "1+1",
+			Instant: kindsys.Ptr(true),
+			RefId:   refID,
+		},
+	}
+	b, _ := json.Marshal(&qm)
+
+	query := backend.DataQuery{
+		RefID: refID,
+		TimeRange: backend.TimeRange{
+			From: time.Unix(1, 0).UTC(),
+			To:   time.Unix(4, 0).UTC(),
+		},
+		JSON: b,
+	}
+	resp, err := i.queryData.Execute(ctx, &backend.QueryDataRequest{
+		PluginContext: req.PluginContext,
+		Queries:       []backend.DataQuery{query},
+	})
+
+	if err != nil {
+		return getHealthCheckMessage(logger, "Your configuration has been saved but there is an error.", err)
```

Comment:
Since this part doesn't do any saving, and just health check, I think it should just be an error message regarding the health check.